### PR TITLE
Reverted the inner loop of RSA keygen to retry when TOO_MANY_ITERATIONS

### PR DESCRIPTION
**Pull request by nebeid**
### Issues:
Addresses CryptoAlg-908

### Description of changes: 
In the FIPS feedback of the previous commit, it was pointed out a flaw in the logic:
Failing for other reasons than TOO_MANY_ITERATIONS would not allow more retries.
So the original loop in `RSA_generate_key_ex_maybe_fips()` in rsa_impl.c as in 6bdd4c3 is brought back to focus on only retrying keygen when TOO_MANY_ITERATIONS is the reason for failure.

An outer loop is added to test for PCT and have its own number of attempts which covers both keygen and PCT failures. This behaviour now matches that of `EC_KEY_generate_key_fips()`.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: crypto/fipsmodule/rsa/rsa_impl.c [hunk 1]
```diff
@@ -1327,37 +1327,46 @@ static int RSA_generate_key_ex_maybe_fips(RSA *rsa, int bits,
...
-           ERR_GET_REASON(err) == RSA_R_TOO_MANY_ITERATIONS);
+    // Perform PCT test in the case of FIPS
+    if (tmp) {
+      if (check_fips && !RSA_check_fips(tmp)) {
+        RSA_free(tmp);
+        tmp = NULL;
+      }
+    }
+    num_attempts++;
+  } while ((num_attempts < MAX_KEYGEN_ATTEMPTS) && (tmp == NULL));
```
Comment (torben-hansen) [inline crypto/fipsmodule/rsa/rsa_impl.c:1369]: `tmp != NULL` in almost-all cases. So check that first I guess

File: crypto/rsa_extra/rsa_test.cc [hunk 2]
```diff
@@ -1082,7 +1086,7 @@ TEST(RSADeathTest, KeygenFailOnceAndDie) {
                [](int event, int n, BN_GENCB *cb_ptr) -> int {
                  bool *failed_ptr = static_cast<bool *>(cb_ptr->arg);
                  if (*failed_ptr) {
-                   ADD_FAILURE() << "Callback called multiple times.";
+                     //ADD_FAILURE() << "Callback called multiple times.";
```
Comment (torben-hansen) [inline crypto/rsa_extra/rsa_test.cc:1089]: Meant to remove?

File: crypto/rsa_extra/rsa_test.cc [hunk 3]
```diff
@@ -1071,7 +1071,11 @@ TEST(RSADeathTest, KeygenFailAndDie) {
   EXPECT_NE(Bytes(der, der_len), Bytes(der3, der3_len));
 }
 
-TEST(RSADeathTest, KeygenFailOnceAndDie) {
+// This is not a death test. It's similar to the test in the non-FIPS case,
+// but because the callback gets called again in the outer loop,
+// this cause of failure is removed.
```
Comment (torben-hansen) [inline crypto/rsa_extra/rsa_test.cc:1076]: re-write sentence.

File: crypto/fipsmodule/rsa/rsa_impl.c [hunk 4]
```diff
@@ -1327,37 +1327,46 @@ static int RSA_generate_key_ex_maybe_fips(RSA *rsa, int bits,
...
   // and thus results in unnecessary complexity.
-  int failures = 0;
-  do {
-    ERR_clear_error();
-    // Generate into scratch space, to avoid leaving partial work on failure.
-    tmp = RSA_new();
-    if (tmp == NULL) {
-      goto out;
-    }
+  failures = 0;
```
Comment (dkostic) [inline crypto/fipsmodule/rsa/rsa_impl.c:1338]: Nit: indentation

File: crypto/fipsmodule/rsa/rsa_impl.c [hunk 5]
```diff
@@ -1327,37 +1327,46 @@ static int RSA_generate_key_ex_maybe_fips(RSA *rsa, int bits,
...
-    RSA_free(tmp);
-    tmp = NULL;
+      err = ERR_peek_error();
+      RSA_free(tmp);
+      tmp = NULL;
+      failures++;
+
+      // Only retry on |RSA_R_TOO_MANY_ITERATIONS|. This is so a caller-induced
+      // failure in |BN_GENCB_call| is still fatal.
+    } while (failures < 4 && ERR_GET_LIB(err) == ERR_LIB_RSA &&
```
Comment (dkostic) [inline crypto/fipsmodule/rsa/rsa_impl.c:1358]: any particular reason to replace MAX_RSA_KEYGEN_ATTEMPTS with 4?

**Review by dkostic — APPROVED** [review]

**Review by andrewhop — APPROVED** [review]

File: crypto/fipsmodule/rsa/rsa_impl.c (hunk 5 above)
Comment (nebeid) [reply crypto/fipsmodule/rsa/rsa_impl.c:1358]: 4 was originally there, and it makes the probability of failure of `rsa_generate_key_impl` down from 2^{-20} to 2^{-80} as per the original comment. The inner loop is exactly as it was before the PCT iterate-and-abort changes and I think of it as one step of RSA key generation.

**Pull request merged**